src/selfplay/loop.cc

```diff
@@ -979,39 +927,23 @@ void ProcessFile(const std::string& file, SyzygyTablebase* tablebase,
             break;
           }
         }
-        int activeZ = ResultForData(fileContents.back());
+        float activeZ[3] = { fileContents.back().result_q,
+                            fileContents.back().result_d,
+                            fileContents.back().plies_left };
         bool deblunderingStarted = false;
         while (true) {
-          if (history.GetLength() == fileContents.size()) {
-            // Game doesn't get to TB, so we need to check if final position is
-            // a blunder.
-            auto& last = fileContents.back();
-            if (last.best_q - last.result_q >
-                deblunderQLastMoveBlunderThreshold) {
-              activeZ = SelectNewZ(Random::Get().GetFloat(1.0), last.best_q,
-                                   last.best_d);
-              deblunderingStarted = true;
-            }
-          } else {
-            auto played = moves[history.GetLength() - 1];
-            auto& cur = fileContents[history.GetLength() - 1];
-            float max_policy = *std::max_element(std::begin(cur.probabilities),
-                                                 std::end(cur.probabilities));
-            int transform = TransformForPosition(input_format, history);
-            int prob_index = played.as_nn_index(transform);
-            float move_policy = cur.probabilities[prob_index];
-            if (move_policy <= deblunderPolicyStrictCutoff * max_policy) {
-              activeZ = SelectNewZ(Random::Get().GetFloat(1.0), cur.best_q,
-                                   cur.best_d);
-              deblunderingStarted = true;
-            } else if (move_policy <= deblunderPolicyWeakCutoff * max_policy) {
-              float q_after = -fileContents[history.GetLength()].best_q;
-              if (cur.best_q - q_after > deblunderQBlunderThreshold) {
-                activeZ = SelectNewZ(Random::Get().GetFloat(1.0), cur.best_q,
-                                     cur.best_d);
-                deblunderingStarted = true;
-              }
-            }
+          auto& cur = fileContents[history.GetLength() - 1];
+          // A blunder is defined by the played move being worse than the
+          // best move by a defined threshold, missing a forced win, or
+          // playing into a proven loss without being forced.
+          if ((cur.best_q - cur.played_q >
+              deblunderQBlunderThreshold) ||
+              (cur.best_q > -1 && cur.played_q < 1 &&
```

Comment:
This check is not 'strictly' true - the cur.best_q == 1 check could be strengthened by using the invariance_info best_q is a proof bit - played_q doesn't have a similar bit we can check - but maybe at least fixing the former wouldn't hurt.

Reply:
so throw in an additional check that `best_q` is a proof (so together with the eval it's a proven win)?

Reply:
yes replace cur.best_q == 1 with something like ((cur.invariance_info & 8 != 0) && cur.best_q == 1)

Reply:
Done.